zebra-chain/src/orchard/keys.rs

```diff
@@ -0,0 +1,1009 @@
+    }
+}
+
+impl From<[u8; 32]> for IncomingViewingKey {
+    /// Generate an _IncomingViewingKey_ from existing bytes.
+    fn from(bytes: [u8; 32]) -> Self {
+        Self {
+            // TODO: handle setting the Network better.
+            network: Network::default(),
+            scalar: pallas::Scalar::from_bytes(&bytes).unwrap(),
+        }
+    }
+}
+
+impl From<FullViewingKey> for IncomingViewingKey {
+    /// Commit^ivk_rivk(ak, nk) :=
+    ///    SinsemillaShortCommit_rcm (︁"z.cash:Orchard-CommitIvk", I2LEBSP_l(ak) || I2LEBSP_l(nk)︁) mod r_P
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#orchardkeycomponents
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+    #[allow(non_snake_case)]
+    fn from(fvk: FullViewingKey) -> Self {
+        let mut M: BitVec<Lsb0, u8> = BitVec::new();
+
+        M.append(&mut BitVec::<Lsb0, u8>::from_slice(
+            &<[u8; 32]>::from(fvk.spend_validating_key)[..],
+        ));
+        M.append(&mut BitVec::<Lsb0, u8>::from_slice(
+            &<[u8; 32]>::from(fvk.nullifier_deriving_key)[..],
+        ));
+
+        // Commit^ivk_rivk
+        let commit_x = sinsemilla_short_commit(
+            fvk.ivk_commit_randomness.into(),
+            b"z.cash:Orchard-CommitIvk",
+            &M,
+        );
+
+        Self {
+            // TODO: handle the network better, maybe an enum variant constraint?
+            network: Network::default(),
+            // mod r_P
+            scalar: pallas::Scalar::from_bytes(&commit_x.into()).unwrap(),
+        }
+    }
+}
+
+impl FromStr for IncomingViewingKey {
+    type Err = SerializationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match bech32::decode(s) {
+            Ok((hrp, bytes, Variant::Bech32)) => {
+                let decoded = Vec::<u8>::from_base32(&bytes).unwrap();
+
+                let mut scalar_bytes = [0u8; 32];
+                scalar_bytes[..].copy_from_slice(&decoded[0..32]);
+
+                Ok(IncomingViewingKey {
+                    network: match hrp.as_str() {
+                        ivk_hrp::MAINNET => Network::Mainnet,
+                        _ => Network::Testnet,
+                    },
+                    scalar: pallas::Scalar::from_bytes(&scalar_bytes).unwrap(),
+                })
+            }
+            _ => Err(SerializationError::Parse("bech32 decoding error")),
+        }
+    }
+}
+
+impl PartialEq<[u8; 32]> for IncomingViewingKey {
+    fn eq(&self, other: &[u8; 32]) -> bool {
+        self.scalar.to_bytes() == *other
+    }
+}
+
+/// Magic human-readable strings used to identify what networks Orchard full
+/// viewing keys are associated with when encoded/decoded with bech32.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#orchardfullviewingkeyencoding
+mod fvk_hrp {
+    pub const MAINNET: &str = "zviewo";
+    pub const TESTNET: &str = "zviewtestorchard";
+}
```

Comment:
Orchard Unified Addresses

Reply:
At present these encodings are defined in https://zips.z.cash/protocol/nu5.pdf, if they are impacted by Unified keys we can remove/rework SpendingKey/IncomingViewingKey/FullViewingKey Display/FromStr accordingly.